[Clang][LoongArch] Support target attribute for function

Ami-zhang opened the pull request:
This adds support under LoongArch for the target("..") attributes.

The supported formats are:
- "arch=<arch>" strings, that specify the architecture features for a function as per the -march=arch option.
- "tune=<cpu>" strings, that specify the tune-cpu cpu for a function as per -mtune.
- "<feature>", "no-<feature>" enabled/disables the specific feature.

Commit:
[Clang][LoongArch] Support target attribute for function

This adds support under LoongArch for the target("..") attributes.

The supported formats are:
- "arch=<arch>" strings, that specify the architecture features for a
  function as per the -march=arch option.
- "tune=<cpu>" strings, that specify the tune-cpu cpu for a function as
  per -mtune.
- "<feature>", "no-<feature>" enabled/disables the specific feature.

tangaac commented on clang/lib/Basic/Targets/LoongArch.cpp:
```diff
@@ -388,6 +388,51 @@ bool LoongArchTargetInfo::handleTargetFeatures(
+  for (auto &Feature : AttrFeatures) {
+    Feature = Feature.trim();
+
+    if (Feature.starts_with("arch=")) {
```
It's recommended to use use a helper function to handle the types of features. Like below.
If it's ok to define a FeatureKind enum, that would be even better.
~~~c++
static std::pair<llvm::StringRef, llvm::StringRef> getFeatureTypeAndValue(llvm::StringRef Feature) {
  auto Split = Feature.split("=");
  if (!Split.second.empty())
    return {Split.first.trim(), Split.second.trim()};
  if (Feature.starts_with("no-"))
    return {"no", Feature.drop_front(3).trim()};
  return {"feature", Feature.trim()};
}
~~~

~~~c++
  for (auto &Feature : AttrFeatures) {
    Feature = Feature.trim();
    auto [Kind, Value] = getFeatureTypeAndValue(Feature);

    if (Kind == "arch") {
    } else if (Kind == "tune") {
    } ..
}
~~~

Ami-zhang replied:
Thanks for the suggestion! Using a helper function with a `FeatureKind` enum is indeed a cleaner and more modular approach. 
However, since `parseTargetAttr` is a common-layer interface that we are simply overriding, I'd prefer to maintain consistency with the common-layer and current RISCV/AArch64 implementations. Keeping the same style offers several benefits:
1. Codebase uniformity – Ensures architectural implementations remain predictable.
2. Easier maintenance – Developers working across architectures can recognize the pattern immediately.
3. Historical consistency – Follows the established patterns for this interface.

If we decide to refactor this interface globally(e.g., by introducing helper functions), I'd be happy to align the LoongArch implementation accordingly.

tangaac replied:
I believe that the previous approach was essentially off-the-cuff, done whichever way was most convenient.

If someone were to implement the same functionality for other architectures as well, they would find three pieces of code with very poor style.

To leave behind code that is more ergonomic for future developers, at the very least, the logic for handling features should be separated and handled individually.

```c++
auto handleArch = [&](StringRef Feature) { ... }
auto handleTune = [&](StringRef Feature) { ... }

if(feature.startwith("arch=") handleArch(feature)
else (feature.startwith("tune=") handleTune(feature)
else if ( ... ) ...
else {

}

```

SixWeining commented on clang/test/Sema/attr-target-loongarch.c:
```diff
@@ -0,0 +1,19 @@
+// RUN: %clang_cc1 -triple loongarch64-linux-gnu  -fsyntax-only -verify %s
+
+// expected-error@+1 {{function multiversioning is not supported on the current target}}
+void __attribute__((target("default"))) bar(void){}
```
```suggestion
void __attribute__((target("default"))) bar(void) {}
```

Ami-zhang replied:
Done.

SixWeining commented on clang/test/Sema/attr-target-loongarch.c:
```diff
@@ -0,0 +1,19 @@
+// expected-error@+1 {{target(arch=..) attribute is not supported on targets missing invalid; specify an appropriate -march= or -mcpu=}}
+void __attribute__((target("arch=invalid"))) foo(void){}
+
+//expected-warning@+1 {{unsupported 'aaa' in the 'target' attribute string; 'target' attribute ignored}}
```
```suggestion
// expected-warning@+1 {{unsupported 'aaa' in the 'target' attribute string; 'target' attribute ignored}}
```

Ami-zhang replied:
Done.

SixWeining commented on clang/test/Sema/attr-target-loongarch.c:
```diff
@@ -0,0 +1,19 @@
+//expected-warning@+1 {{unsupported 'aaa' in the 'target' attribute string; 'target' attribute ignored}}
+int __attribute__((target("aaa"))) test_feature(void) { return 4; }
+
+//expected-warning@+1 {{unsupported 'aaa' in the 'target' attribute string; 'target' attribute ignored}}
```
```suggestion
// expected-warning@+1 {{unsupported 'aaa' in the 'target' attribute string; 'target' attribute ignored}}
```

Commit:
[Loongarch] Refactor target attribute parsing

Introduce AttrFeatureKind enum to classify target attributes, and
implement getAttrFeatureTypeAndValue() helper for centralized parsing.

heiher commented on clang/lib/Basic/Targets/LoongArch.cpp:
```diff
@@ -388,6 +388,76 @@ bool LoongArchTargetInfo::handleTargetFeatures(
+  }
+  if (AttrFeature.starts_with("no-"))
+    return {AttrFeatureKind::NoFeature, AttrFeature.drop_front(3).trim()};
+  return {AttrFeatureKind::Feature, AttrFeature.trim()};
```
The caller has already  performed `trim()`, so it's unnecessary here, just like `starts_with()` above doesn't use it either.

```suggestion
  return {AttrFeatureKind::Feature, AttrFeature};
```

heiher reviewed (approved):
LGTM with nits.

Ami-zhang commented on clang/test/CodeGen/LoongArch/targetattr.c:
```diff
@@ -0,0 +1,92 @@
+// CHECK-NEXT:  [[ENTRY:.*:]]
+// CHECK-NEXT:    ret void
+//
+void tuneLa664() {}
```
Done.

Ami-zhang commented on llvm/include/llvm/TargetParser/LoongArchTargetParser.h:
```diff
@@ -85,6 +85,7 @@ struct ArchInfo {
 };
 
 bool isValidArchName(StringRef Arch);
+bool isValidFeatureName(StringRef Arch);
```
Done.

wangleiat commented on llvm/lib/TargetParser/LoongArchTargetParser.cpp:
```diff
@@ -34,6 +34,19 @@ bool LoongArch::isValidArchName(StringRef Arch) {
+bool LoongArch::isValidFeatureName(StringRef Feature) {
+  if (Feature.starts_with("+") || Feature.starts_with("-")) {
+    return false;
+  }
```
The braces can be omitted.

Ami-zhang replied:
Done.

wangleiat commented on llvm/lib/TargetParser/LoongArchTargetParser.cpp:
```diff
@@ -34,6 +34,19 @@ bool LoongArch::isValidArchName(StringRef Arch) {
+  if (Feature.starts_with("+") || Feature.starts_with("-")) {
+    return false;
+  }
+  for (const auto F : AllFeatures) {
+    StringRef CanonicalName =
```
Beware unnecessary copies.
`auto &F`

Ami-zhang replied:
Done.

wangleiat reviewed (approved):
LGTM with some minor nits, thanks.

Ami-zhang commented on clang/lib/Basic/Targets/LoongArch.cpp:
```diff
@@ -388,6 +388,76 @@ bool LoongArchTargetInfo::handleTargetFeatures(
   return true;
 }
 
+enum class AttrFeatureKind { Arch, Tune, NoFeature, Feature, Invalid };
```
Removed.
